src/test/java/com/google/jenkins/plugins/k8sengine/KubernetesEngineBuilderTest.java

```diff
@@ -655,53 +656,116 @@ public void testDoCheckClusterNameOkWithValidInputs() {
+  public void testDoCheckClusterNameMessageWithIOException() throws IOException {
+    DescriptorImpl descriptor = setUpClusterDescriptor(ImmutableList.of(), null, new IOException());
     FormValidation result =
         descriptor.doCheckClusterName(
-            jenkins, TEST_CLUSTER, TEST_CREDENTIALS_ID, ERROR_PROJECT_ID, TEST_ZONE_A);
+            jenkins, TEST_CLUSTER, TEST_CREDENTIALS_ID, TEST_PROJECT_ID, TEST_ZONE_A);
     assertNotNull(result);
     assertEquals(Messages.KubernetesEngineBuilder_ClusterVerificationError(), result.getMessage());
   }
 
   @Test
-  public void testDoCheckClusterNameMessageWithAbortExceptionAndEmptyClusterName() {
+  public void testDoCheckClusterNameMessageWithAbortExceptionAndEmptyClusterName()
+      throws IOException {
+    DescriptorImpl descriptor =
+        setUpClusterDescriptor(ImmutableList.of(), new AbortException(), null);
     FormValidation result =
         descriptor.doCheckClusterName(
-            jenkins, null, ERROR_CREDENTIALS_ID, TEST_PROJECT_ID, TEST_ZONE_A);
+            jenkins, null, TEST_CREDENTIALS_ID, TEST_PROJECT_ID, TEST_ZONE_A);
     assertNotNull(result);
     assertEquals(Messages.KubernetesEngineBuilder_CredentialAuthFailed(), result.getMessage());
   }
 
   @Test
-  public void testDoCheckClusterNameMessageWithIOExceptionAndEmptyClusterName() {
+  public void testDoCheckClusterNameMessageWithIOExceptionAndEmptyClusterName() throws IOException {
+    DescriptorImpl descriptor = setUpClusterDescriptor(ImmutableList.of(), null, new IOException());
     FormValidation result =
         descriptor.doCheckClusterName(
-            jenkins, null, TEST_CREDENTIALS_ID, ERROR_PROJECT_ID, TEST_ZONE_A);
+            jenkins, null, TEST_CREDENTIALS_ID, TEST_PROJECT_ID, TEST_ZONE_A);
     assertNotNull(result);
     assertEquals(Messages.KubernetesEngineBuilder_ClusterVerificationError(), result.getMessage());
   }
 
-  private static void initProjects(List<String> projectNames) {
-    projectNames.forEach(p -> listOfProjects.add(new Project().setProjectId(p)));
+  private DescriptorImpl setUpDescriptor(AbortException abortException) throws AbortException {
+    DescriptorImpl descriptor = Mockito.spy(DescriptorImpl.class);
+    if (abortException != null) {
+      Mockito.doThrow(abortException)
+          .when(descriptor)
+          .getClientFactory(any(Jenkins.class), anyString());
+      return descriptor;
+    }
+    Mockito.doReturn(clientFactory)
+        .when(descriptor)
+        .getClientFactory(any(Jenkins.class), anyString());
+
+    return descriptor;
+  }
+
+  private DescriptorImpl setUpProjectDescriptor(
+      List<String> initialProjects,
+      String defaultProjectId,
+      AbortException abortException,
+      IOException ioException)
+      throws IOException {
+    DescriptorImpl descriptor = setUpDescriptor(abortException);
+
+    Mockito.when(clientFactory.getDefaultProjectId()).thenReturn(defaultProjectId);
```

Comment:
Unfortunately, we're still in a situation where shared state is being altered here. Because the clientFactory and clients spies are all instance variables, the tests executing in parallel will all be changing their state concurrently. These all need to be changed to locally scoped for each test. Same feedback for the other test classes as well.

Reply:
Makes sense, I'll change these, and the other test classes will be fixed in the next set of commits I submit.